file/file_prefetch_buffer.cc

```diff
@@ -399,289 +382,291 @@ void FilePrefetchBuffer::ReadAheadSizeTuning(
   //      already full.
   //    - If second buffer is empty, it will go for ReadAsync for second buffer.
-  if (!bufs_[curr_].async_read_in_progress_ && DoesBufferContainData(curr_) &&
-      IsDataBlockInBuffer(offset, length, curr_)) {
-    // Whole data is in curr_.
-    UpdateBuffersIfNeeded(offset, length);
-    if (!IsSecondBuffEligibleForPrefetching()) {
+  if (!buf->async_read_in_progress_ && buf->DoesBufferContainData() &&
+      buf->IsDataBlockInBuffer(offset, length)) {
+    // Whole data is in buffer.
+    if (!IsEligibleForFurtherPrefetching()) {
       UpdateStats(/*found_in_buffer=*/true, original_length);
       return s;
     }
   } else {
-    // After poll request, curr_ might be empty because of IOError in
-    // callback while reading or may contain required data.
-    PollAndUpdateBuffersIfNeeded(offset, length);
+    // NOTE - After this poll request, first buffer might be empty because of
+    // IOError in callback while reading or it may contains the required data.
+    PollIfNeeded(offset, length);
   }
 
-  if (copy_to_third_buffer) {
+  if (copy_to_overlap_buffer) {
     offset = tmp_offset;
     length = tmp_length;
   }
 
-  // 4. After polling and swapping buffers, if all the requested bytes are in
-  // curr_, it will only go for async prefetching.
-  // copy_to_third_buffer is a special case so it will be handled separately.
-  if (!copy_to_third_buffer && DoesBufferContainData(curr_) &&
-      IsDataBlockInBuffer(offset, length, curr_)) {
-    offset += length;
-    length = 0;
-
-    // Since async request was submitted directly by calling PrefetchAsync in
-    // last call, we don't need to prefetch further as this call is to poll
-    // the data submitted in previous call.
-    if (explicit_prefetch_submitted_) {
-      return s;
-    }
-    if (!IsSecondBuffEligibleForPrefetching()) {
-      UpdateStats(/*found_in_buffer=*/true, original_length);
-      return s;
+  // After polling, if all the requested bytes are in first buffer, it will only
+  // go for async prefetching.
+  if (buf->DoesBufferContainData()) {
+    if (copy_to_overlap_buffer) {
+      // Data is overlapping i.e. some of the data has been copied to overlap
+      // buffer and remaining will be updated below.
+      size_t initial_buf_size = overlap_buf_->CurrentSize();
+      CopyDataToBuffer(buf, offset, length);
+      UpdateStats(
+          /*found_in_buffer=*/false,
+          overlap_buf_->CurrentSize() - initial_buf_size);
+
+      // Length == 0: All the requested data has been copied to overlap buffer
+      // and it has already gone for async prefetching. It can return without
+      // doing anything further. Length > 0: More data needs to be consumed so
+      // it will continue async and sync prefetching and copy the remaining data
+      // to overlap buffer in the end.
+      if (length == 0) {
+        UpdateStats(/*found_in_buffer=*/true, length);
+        return s;
+      }
```

Comment:
I wonder if this belongs in `HandleOverlappingData()`. It's hard to say where that function should end and this one picks up. If we say logic specific to overlapping buffers belongs there, then we should probably move this.